### ISSUE TITLE
DDIC searchhelp generator

### ISSUE THREAD

**Issue by axelmohnen** [abap2UI5/abap2UI5#507]
Hi oblomov,

I had the requirement to reuse standard and custom DDIC searchhelps in my ABAP2UI5 app.
Therefore I developed an "DDIC sreachhelp generator".
This is just a first version and for sure it missed a lot of functionalities, but I guess it's a good kick-off and a opener for discussion.
Basically its just a simple static method "GENERATE_DDIC_SHLP" to be called in order to generate all the search fields and the result table of a corresponding DDIC searchhelp.
You only need to declare a public structure for the searchhelp input-fields and a public internal table for the searchhelp result in your ABAP2UI5 application class. 

Class: ZCL_ABAP2UI5_UTIL
`
class ZCL_ABAP2UI5_UTIL definition
  public
  final
  create public .

public section.

  class-methods GENERATE_DDIC_SHLP
    importing
      !IR_PARENT type ref to Z2UI5_CL_XML_VIEW
      !IR_CLIENT type ref to Z2UI5_IF_CLIENT
      !IR_CONTROLLER type ref to OBJECT
      !IV_SHLP_ID type SHLPNAME
      !IV_RESULT_ITAB_NAME type CLIKE
      !IV_RESULT_ITAB_EVENT type CLIKE
      !IV_SHLP_FIELDS_STRUC_NAME type CLIKE .
protected section.
private section.
ENDCLASS.

CLASS ZCL_ABAP2UI5_UTIL IMPLEMENTATION.

* <SIGNATURE>---------------------------------------------------------------------------------------+
* | Static Public Method ZCL_ABAP2UI5_UTIL=>GENERATE_DDIC_SHLP
* +-------------------------------------------------------------------------------------------------+
* | [--->] IR_PARENT                      TYPE REF TO Z2UI5_CL_XML_VIEW
* | [--->] IR_CLIENT                      TYPE REF TO Z2UI5_IF_CLIENT
* | [--->] IR_CONTROLLER                  TYPE REF TO OBJECT
* | [--->] IV_SHLP_ID                     TYPE        SHLPNAME
* | [--->] IV_RESULT_ITAB_NAME            TYPE        CLIKE
* | [--->] IV_RESULT_ITAB_EVENT           TYPE        CLIKE
* | [--->] IV_SHLP_FIELDS_STRUC_NAME      TYPE        CLIKE
* +--------------------------------------------------------------------------------------</SIGNATURE>
  METHOD generate_ddic_shlp.
*----------------------------------------------------------------------*
* LOCAL DATA DEFINITION
*----------------------------------------------------------------------*
    DATA: ls_shlp             TYPE  shlp_descr,
          lv_grid_form_no     TYPE i,
          lt_arg              TYPE TABLE OF stringval,
          lv_arg_fieldname    TYPE stringval,
          lv_cell_fieldname   TYPE stringval,
          lv_path_result_itab TYPE stringval,
          lv_path_shlp_fields TYPE stringval.

    FIELD-SYMBOLS: <ls_fielddescr>  TYPE dfies,
                   <ls_fieldprop>   TYPE ddshfprop,
                   <lt_result_itab> TYPE ANY TABLE,
                   <ls_shlp_fields> TYPE any,
                   <lv_field>       TYPE any.

* ---------- Get result itab reference ------------------------------------------------------------
    lv_path_result_itab = 'IR_CONTROLLER->' && iv_result_itab_name.
    ASSIGN (lv_path_result_itab) TO <lt_result_itab>.

* ---------- Get searchhelp input fields structure reference --------------------------------------
    lv_path_shlp_fields = 'IR_CONTROLLER->' && iv_shlp_fields_struc_name.
    ASSIGN (lv_path_shlp_fields) TO <ls_shlp_fields>.

    IF <lt_result_itab> IS NOT ASSIGNED OR
      <ls_shlp_fields> IS NOT ASSIGNED.
      RETURN.
    ENDIF.

* -------------------------------------------------------------------------------------------------
* Searchfield Grid
* -------------------------------------------------------------------------------------------------
    DATA(lr_grid_shlp) = ir_parent->grid( 'L3 M3 S3' )->content( 'layout' ).

* ---------- Create 4 forms (grid columns) --------------------------------------------------------
    DATA(lr_form_shlp_1) = lr_grid_shlp->simple_form( )->content( 'form' ).
    DATA(lr_form_shlp_2) = lr_grid_shlp->simple_form( )->content( 'form' ).
    DATA(lr_form_shlp_3) = lr_grid_shlp->simple_form( )->content( 'form' ).
    DATA(lr_form_shlp_4) = lr_grid_shlp->simple_form( )->content( 'form' ).

* ---------- Get searchhelp description -----------------------------------------------------------
    CALL FUNCTION 'F4IF_GET_SHLP_DESCR'
      EXPORTING
        shlpname = iv_shlp_id
      IMPORTING
        shlp     = ls_shlp.

    LOOP AT ls_shlp-fielddescr ASSIGNING <ls_fielddescr>.
* ---------- Init loop data -----------------------------------------------------------------------
      UNASSIGN: <lv_field>.

* ---------- Get field reference ------------------------------------------------------------------
      ASSIGN COMPONENT <ls_fielddescr>-fieldname OF STRUCTURE <ls_shlp_fields> TO <lv_field>.
      IF <lv_field> IS NOT ASSIGNED.
        CONTINUE.
      ENDIF.

* ---------- Determine grid form number -----------------------------------------------------------
      IF lv_grid_form_no IS INITIAL.
        lv_grid_form_no = 1.
      ELSEIF lv_grid_form_no = 4.
        lv_grid_form_no = 1.
      ELSE.
        lv_grid_form_no = lv_grid_form_no + 1.
      ENDIF.

      CASE lv_grid_form_no.
        WHEN 1.
* ---------- Grid 1--------------------------------------------------------------------------------
* ---------- Set field label ----------------------------------------------------------------------
          lr_form_shlp_1->label( <ls_fielddescr>-scrtext_l ).

* ---------- Set input field ----------------------------------------------------------------------
          CASE <ls_fielddescr>-datatype.
            WHEN 'DATS'.
              lr_form_shlp_1->date_picker( value  = ir_client->_bind_edit( <lv_field> ) ).
            WHEN 'TIMS'.
              lr_form_shlp_1->time_picker( value  = ir_client->_bind_edit( <lv_field> ) ).
            WHEN OTHERS.
              lr_form_shlp_1->input( value = ir_client->_bind_edit( <lv_field> ) ).
          ENDCASE.

        WHEN 2.
* ---------- Grid 2--------------------------------------------------------------------------------
* ---------- Set field label ----------------------------------------------------------------------
          lr_form_shlp_2->label( zcl_gu=>get_data_element_labels( iv_data_element = <ls_fielddescr>-rollname ) ).

* ---------- Set input field ----------------------------------------------------------------------
          CASE <ls_fielddescr>-datatype.
            WHEN 'DATS'.
              lr_form_shlp_2->date_picker( value  = ir_client->_bind_edit( <lv_field> ) ).
            WHEN 'TIMS'.
              lr_form_shlp_2->time_picker( value  = ir_client->_bind_edit( <lv_field> ) ).
            WHEN OTHERS.
              lr_form_shlp_2->input( value = ir_client->_bind_edit( <lv_field> ) ).
          ENDCASE.

        WHEN 3.
* ---------- Grid 3--------------------------------------------------------------------------------
* ---------- Set field label ----------------------------------------------------------------------
          lr_form_shlp_3->label( zcl_gu=>get_data_element_labels( iv_data_element = <ls_fielddescr>-rollname ) ).

* ---------- Set input field ----------------------------------------------------------------------
          CASE <ls_fielddescr>-datatype.
            WHEN 'DATS'.
              lr_form_shlp_3->date_picker( value  = ir_client->_bind_edit( <lv_field> ) ).
            WHEN 'TIMS'.
              lr_form_shlp_3->time_picker( value  = ir_client->_bind_edit( <lv_field> ) ).
            WHEN OTHERS.
              lr_form_shlp_3->input( value = ir_client->_bind_edit( <lv_field> ) ).
          ENDCASE.

        WHEN 4.
* ---------- Grid 4--------------------------------------------------------------------------------
* ---------- Set field label ----------------------------------------------------------------------
          lr_form_shlp_4->label( zcl_gu=>get_data_element_labels( iv_data_element = <ls_fielddescr>-rollname ) ).

* ---------- Set input field ----------------------------------------------------------------------
          CASE <ls_fielddescr>-datatype.
            WHEN 'DATS'.
              lr_form_shlp_4->date_picker( value  = ir_client->_bind_edit( <lv_field> ) ).
            WHEN 'TIMS'.
              lr_form_shlp_4->time_picker( value  = ir_client->_bind_edit( <lv_field> ) ).
            WHEN OTHERS.
              lr_form_shlp_4->input( value = ir_client->_bind_edit( <lv_field> ) ).
          ENDCASE.

      ENDCASE.
    ENDLOOP.

* ---------- Create table -------------------------------------------------------------------------
    DATA(lr_table) = ir_parent->table( items = ir_client->_bind_edit( <lt_result_itab> ) ).
* ---------- Create Columns -----------------------------------------------------------------------
    DATA(lr_columns) = lr_table->columns( ).

* ---------- Set column ---------------------------------------------------------------------------
    LOOP AT ls_shlp-fielddescr ASSIGNING <ls_fielddescr>.
      lr_columns->column( )->text( zcl_gu=>get_data_element_labels( iv_data_element = <ls_fielddescr>-rollname ) ).
    ENDLOOP.

* ---------- Build export parameter list ----------------------------------------------------------
    LOOP AT ls_shlp-fieldprop ASSIGNING <ls_FIELDPROP> WHERE shlpoutput = abap_true.
* ---------- Init loop data -----------------------------------------------------------------------
      CLEAR: lv_arg_fieldname.

* ---------- Build parameter name -----------------------------------------------------------------
      lv_arg_fieldname = `${` && <ls_FIELDPROP>-fieldname && `}`.

* ---------- Collect output fields ----------------------------------------------------------------
      APPEND lv_arg_fieldname TO lt_arg.
    ENDLOOP.

    DATA(lr_item) = lr_table->items(
        )->column_list_item( type = 'Navigation'  press = ir_client->_event( val    = iv_result_itab_event
                                                                             t_arg  = lt_arg ) ).

* ---------- Set cell content ---------------------------------------------------------------------
    LOOP AT ls_shlp-fielddescr ASSIGNING <ls_fielddescr>.
* ---------- Init loop data -----------------------------------------------------------------------
      CLEAR: lv_cell_fieldname.

* ---------- Build cell name ----------------------------------------------------------------------
      lv_cell_fieldname = `{` && <ls_fielddescr>-fieldname && `}`.

      lr_item->cells( )->text( lv_cell_fieldname ).
    ENDLOOP.
  ENDMETHOD.
ENDCLASS.
`

Call of class ZCL_ABAP2UI5:
`
    zcl_cc_abap2ui5_util=>generate_ddic_shlp(
                                              ir_parent                 = [Instance of Z2UI5_CL_XML_VIEW]
                                              ir_client                 = ir_client
                                              ir_controller             = [your main ABAP2UI5 app class]
                                              iv_shlp_id                = [ID of the elementary help from DDIC]
                                              iv_result_itab_name       = [ITAB name for the shlp result]
                                              iv_result_itab_event      = [SHLP result table select event]
                                              iv_shlp_fields_struc_name = [STRUC name of SHLP fields] ).
`

Kind regards
Axel

**Comment by choper725** [comment]
seems to be a good idea.. maybe a custom control in the xml view class might fit smoothly to this project

**Comment by oblomov-dev** [comment]
Cool, thank you for this idea and the work! Having a generic approach for search helps is often needed and can save a lot of time! Do you also have a small demo app to see exactly how it is called? 

Maybe some of your comments made problems on github, thats why the code was not formatted correctly. I adjusted it:
```abap
CLASS z2ui5_tool_cl_app_10 DEFINITION
  PUBLIC
  FINAL
  CREATE PUBLIC .

  PUBLIC SECTION.

    CLASS-METHODS generate_ddic_shlp
      IMPORTING
        !ir_parent                 TYPE REF TO z2ui5_cl_xml_view
        !ir_client                 TYPE REF TO z2ui5_if_client
        !ir_controller             TYPE REF TO object
        !iv_shlp_id                TYPE shlpname
        !iv_result_itab_name       TYPE clike
        !iv_result_itab_event      TYPE clike
        !iv_shlp_fields_struc_name TYPE clike.

  PROTECTED SECTION.
  PRIVATE SECTION.
ENDCLASS.

CLASS z2ui5_tool_cl_app_10 IMPLEMENTATION.

  METHOD generate_ddic_shlp.

    DATA: ls_shlp             TYPE shlp_descr,
          lv_grid_form_no     TYPE i,
          lt_arg              TYPE TABLE OF stringval,
          lv_arg_fieldname    TYPE stringval,
          lv_cell_fieldname   TYPE stringval,
          lv_path_result_itab TYPE stringval,
          lv_path_shlp_fields TYPE stringval.

    FIELD-SYMBOLS: <ls_fielddescr>  TYPE dfies,
                   <ls_fieldprop>   TYPE ddshfprop,
                   <lt_result_itab> TYPE ANY TABLE,
                   <ls_shlp_fields> TYPE any,
                   <lv_field>       TYPE any.


    lv_path_result_itab = 'IR_CONTROLLER->' && iv_result_itab_name.
    ASSIGN (lv_path_result_itab) TO <lt_result_itab>.

    lv_path_shlp_fields = 'IR_CONTROLLER->' && iv_shlp_fields_struc_name.
    ASSIGN (lv_path_shlp_fields) TO <ls_shlp_fields>.

    IF <lt_result_itab> IS NOT ASSIGNED OR
    <ls_shlp_fields> IS NOT ASSIGNED.
      RETURN.
    ENDIF.

    DATA(lr_grid_shlp) = ir_parent->grid( 'L3 M3 S3' )->content( 'layout' ).
    DATA(lr_form_shlp_1) = lr_grid_shlp->simple_form( )->content( 'form' ).
    DATA(lr_form_shlp_2) = lr_grid_shlp->simple_form( )->content( 'form' ).
    DATA(lr_form_shlp_3) = lr_grid_shlp->simple_form( )->content( 'form' ).
    DATA(lr_form_shlp_4) = lr_grid_shlp->simple_form( )->content( 'form' ).

    CALL FUNCTION 'F4IF_GET_SHLP_DESCR'
      EXPORTING
        shlpname = iv_shlp_id
      IMPORTING
        shlp     = ls_shlp.

    LOOP AT ls_shlp-fielddescr ASSIGNING <ls_fielddescr>.

      UNASSIGN: <lv_field>.

      ASSIGN COMPONENT <ls_fielddescr>-fieldname OF STRUCTURE <ls_shlp_fields> TO <lv_field>.
      IF <lv_field> IS NOT ASSIGNED.
        CONTINUE.
      ENDIF.

      IF lv_grid_form_no IS INITIAL.
        lv_grid_form_no = 1.
      ELSEIF lv_grid_form_no = 4.
        lv_grid_form_no = 1.
      ELSE.
        lv_grid_form_no = lv_grid_form_no + 1.
      ENDIF.

      CASE lv_grid_form_no.
        WHEN 1.
          lr_form_shlp_1->label( <ls_fielddescr>-scrtext_l ).

          CASE <ls_fielddescr>-datatype.
            WHEN 'DATS'.
              lr_form_shlp_1->date_picker( value = ir_client->_bind_edit( <lv_field> ) ).
            WHEN 'TIMS'.
              lr_form_shlp_1->time_picker( value = ir_client->_bind_edit( <lv_field> ) ).
            WHEN OTHERS.
              lr_form_shlp_1->input( value = ir_client->_bind_edit( <lv_field> ) ).
          ENDCASE.

        WHEN 2.
          lr_form_shlp_2->label( <ls_fielddescr>-rollname ).

          CASE <ls_fielddescr>-datatype.
            WHEN 'DATS'.
              lr_form_shlp_2->date_picker( value = ir_client->_bind_edit( <lv_field> ) ).
            WHEN 'TIMS'.
              lr_form_shlp_2->time_picker( value = ir_client->_bind_edit( <lv_field> ) ).
            WHEN OTHERS.
              lr_form_shlp_2->input( value = ir_client->_bind_edit( <lv_field> ) ).
          ENDCASE.

        WHEN 3.
          lr_form_shlp_3->label( <ls_fielddescr>-rollname ).

          CASE <ls_fielddescr>-datatype.
            WHEN 'DATS'.
              lr_form_shlp_3->date_picker( value = ir_client->_bind_edit( <lv_field> ) ).
            WHEN 'TIMS'.
              lr_form_shlp_3->time_picker( value = ir_client->_bind_edit( <lv_field> ) ).
            WHEN OTHERS.
              lr_form_shlp_3->input( value = ir_client->_bind_edit( <lv_field> ) ).
          ENDCASE.

        WHEN 4.
          lr_form_shlp_4->label( <ls_fielddescr>-rollname ).

          CASE <ls_fielddescr>-datatype.
            WHEN 'DATS'.
              lr_form_shlp_4->date_picker( value = ir_client->_bind_edit( <lv_field> ) ).
            WHEN 'TIMS'.
              lr_form_shlp_4->time_picker( value = ir_client->_bind_edit( <lv_field> ) ).
            WHEN OTHERS.
              lr_form_shlp_4->input( value = ir_client->_bind_edit( <lv_field> ) ).
          ENDCASE.

      ENDCASE.
    ENDLOOP.

    DATA(lr_table) = ir_parent->table( items = ir_client->_bind_edit( <lt_result_itab> ) ).
    DATA(lr_columns) = lr_table->columns( ).

    LOOP AT ls_shlp-fielddescr ASSIGNING <ls_fielddescr>.
      lr_columns->column( )->text( <ls_fielddescr>-rollname ).
    ENDLOOP.

    LOOP AT ls_shlp-fieldprop ASSIGNING <ls_fieldprop> WHERE shlpoutput = abap_true.

      CLEAR: lv_arg_fieldname.
      lv_arg_fieldname = `${` && <ls_fieldprop>-fieldname && `}`.
      APPEND lv_arg_fieldname TO lt_arg.
    ENDLOOP.

    DATA(lr_item) = lr_table->items(
    )->column_list_item( type = 'Navigation' press = ir_client->_event( val = iv_result_itab_event
    t_arg = lt_arg ) ).

    LOOP AT ls_shlp-fielddescr ASSIGNING <ls_fielddescr>.

      CLEAR: lv_cell_fieldname.
      lv_cell_fieldname = `{` && <ls_fielddescr>-fieldname && `}`.
      lr_item->cells( )->text( lv_cell_fieldname ).
    ENDLOOP.
  ENDMETHOD.
ENDCLASS.
```

Best regards.

**Comment by choper725** [comment]
i tried moving it to the xml view, it worked well, but no compatible on downport due to the fm import export parameters, maybe worth a shot retyping them locally with string datatype.

**Comment by axelmohnen** [comment]
Hi,
I created a new demo class "Z2UI5_CL_APP_DEMO_102" for the DDIC searchhelp generator (See pull request).
I added the two static methods GENERATE_DDIC_SHLP and SELECT_DDIC_SHLP to the class in order to run it standalone.

How shall we implement this into the xml view class as a custom control?

Thanks,
Axel

**Comment by choper725** [comment]
see PR #508  for the class generate_ddic_shlp.

it was working but the error i got in that pr was compatibility issue..

**Comment by choper725** [comment]
something not working right on my end in the demo..
i tried to switch the SH to the se24 SH called SEO_CLASSES_INTERFACES, not getting any results...
here is my code
[demo 102.txt](https://github.com/abap2UI5/abap2UI5/files/12457278/demo.102.txt)

**Comment by axelmohnen** [comment]
@choper725: unfortunately, I can't re-produce your error.
Your example was working out of the box!
![image](https://github.com/abap2UI5/abap2UI5/assets/22371169/9017ed03-2fc1-47c5-8714-f5eff7afdd80)

What is your error message or erroneous behavior?

**Comment by choper725** [comment]
there wasnt an error, just not getting any data back.. the fm suppose to get the data returned empty..

**Comment by oblomov-dev** [comment]
Nice example! Looks great.

@choper725 
The downport is not the main problem, more the compatibility to the two language flavors "cloud" and "standard".

To add this to the xml view directly, i recommend:
* only use basic abap types instead of ddic types
* function modules can be encapsulated in a concept similar to here: https://github.com/abap2UI5/abap2UI5/blob/e8bf4ba9a0b518cd8912d9d4547239fd5f904dc3/src/00/z2ui5_cl_fw_utility.clas.abap#L267

The dynamic call makes it possible to use the fm in an on-premise system and in an cloud system there will be then no result or an exception with an information can be shown.

**Comment by choper725** [comment]
thx @oblomov-dev  for the info.

@axelmohnen  it would be easier if there was a single method for this functionality (inculding getting the data),
just like creating a input and connecting is f4 to a popup view with ")->generate_ddic_shlp" that will handle all the layout and logic..

**Comment by choper725** [comment]
hi @axelmohnen 
see demo 102, i got 1 of the two static methods to be generic,
im trying to male both of them generic, but i dont know how to trigger an event from the view to run the select_ddic method.. any ideas ?

**Comment by axelmohnen** [comment]
Hi @choper725,
yes, that's difficult to bring the rendering part and event handling part together in one generic method.
A simple way would be to use an indicator as import parameter to separate between rendering and event handling.
The same generic method could be call than at rendering and event handling time, just with different indicators.
Thanks,
Axel

**Comment by oblomov-dev** [comment]
I am not deep in this topic, but maybe creating a separated app only for the popup is an option to keep ui+event handling at one place. See an example here:
Z2UI5_CL_DEMO_APP_012

**Comment by axelmohnen** [comment]
Yes, this could work. I will check this....

**Comment by axelmohnen** [comment]
Hi,
I created two new demo apps.
Demo "z2ui5_cl_app_demo_104" represents the DDIC searchhelp generator in separated app.
Demo "z2ui5_cl_app_demo_105" is the started app for the demo.

New generator features:
- Fully dynamically data definition. No need anymore to define the DDIC shlp fields. Only the SHLP ID is needed.
- The searchhelp includes the "select-option" approach using multi-input with tokens.

Open points:
Include DDIC composite searchhelps using the "filter tab bar" control

The demo is available via PR.

Thanks,
Axel

**Comment by oblomov-dev** [comment]
Hu @axelmohnen,

thank you for the work!

Not sure if this is my system but when i try to use the functionality, i get always this error when i press something:
<img width="500" alt="image" src="https://github.com/abap2UI5/abap2UI5/assets/102328295/4957a743-4c9f-4183-b7d2-ed16e4a36250">
<img width="500" alt="image" src="https://github.com/abap2UI5/abap2UI5/assets/102328295/3d167ff3-6f35-4217-be8c-e5b4dbc3d87f">
What do i have to do here?

Added also a small bugfix to abap2UI5, had some problems with the serialization.

Yes good question, will think about it where is the best place to store it. 

Best regards.

**Comment by choper725** [comment]
what is the minimum version if rtti? is it possible its > 7.50 ??

**Comment by oblomov-dev** [comment]
Mhh not sure, but RTTI goes very low, it think this is not the problem. 
But i also did not debug the error by myself, just wondering if you get the same error?

**Comment by choper725** [comment]
> Mhh not sure, but RTTI goes very low, it think this is not the problem.
> But i also did not debug the error by myself, just wondering if you get the same error?

actually im just asking a general question as this is implemented as a feature in this project

**Comment by oblomov-dev** [comment]
Do you mean RTTI or S-RTTI?

Seems that the Searchhelp is created with an older abap2UI5 version, the binding logic changed slightly. Added a new error message and now i get this:
<img width="418" alt="image" src="https://github.com/abap2UI5/abap2UI5/assets/102328295/f48e70e1-2f8d-4ed1-a3e3-027d5804af13">
Seems that some data is not available anymore in the end of the request when the response is created.

**Comment by choper725** [comment]
oh sorry i mean S-RTTI

**Comment by axelmohnen** [comment]
@oblomov-dev:
I'm using the ABAP2UI5 version v.1.25.2.
Do you have the latest version of S-RTTI installed?
The "MR_SHLP_FIELDS" data type will be defined in method "BUILD_DATA_REF" of class "Z2UI5_CL_APP_DEMO_104".
Could you please set a break-point at the very end of the method and check that the data references MR_SHLP_FIELDS and MR_SHLP_RESULT" are correctly created.

**Comment by oblomov-dev** [comment]
Yes the binding changed slightly since this tag. I changed now the binding _bind to _bind_local at some positions and updated the demo. Now it seems to work better:
<img width="600" alt="image" src="https://github.com/abap2UI5/abap2UI5/assets/102328295/672fb446-62e0-4d19-9cc0-29e5d8caf7be">
<img width="600" alt="image" src="https://github.com/abap2UI5/abap2UI5/assets/102328295/12ec080d-f3e7-4fb6-a733-d776a0aec83a">
Which values are you using for testing? Everything is still a bit empty on my side.

**Comment by axelmohnen** [comment]
I'm using the DDIC Shlp "BUPAP". That means you need some business partners in your system (Table BUT000).
You can switch to any DDIC SHLP, by changing the ID in demo "105".
You have to "press" the "GO" button to select data. You can do this without any filter criteria.

**Comment by oblomov-dev** [comment]
Ah ok now it works! Great job!

**Comment by oblomov-dev** [comment]
yes every system needs the installation of s-rrti to use this. But S-RTTI is also compatible to low releases, so this should be no problem.

**Comment by axelmohnen** [comment]
In my first tests I received the error message which was pointing me to the S-RTTI project.

**Comment by choper725** [comment]
s-rtti not compatible to 7.50 sp13, due to cl_abap_enumdescr class not existing

**Comment by oblomov-dev** [comment]
yes because this releases also have no enums. you can just delete this class and it will run again i think. As far as i know S-RTTI is just a copy of the normal RTTI classes with some small adjustments making them serializable.

**Comment by choper725** [comment]
> yes because this releases also have no enums. you can just delete this class and it will run again i think. As far as i know S-RTTI is just a copy of the normal RTTI classes with some small adjustments making them serializable.

this one ? zcl_srtti_elemdescr

**Comment by oblomov-dev** [comment]
yes and if there are references in other classes to this, just make a comment out of it.

**Comment by choper725** [comment]
thanks ill try !

**Comment by choper725** [comment]
got it to work by,

1. deleting class, zcl_srtti_ebumdescr

2. commenting:
![image](https://github.com/abap2UI5/abap2UI5/assets/129428114/7ec58e67-1d54-4c1e-a263-b1cd09a0e7d5)

in class: zcl_srtti_typedescr

thanks @oblomov-dev 

BTW:
1. seems to not work when input filter data from SH field only, and not working like previously when typing into the input control
2. when opening the main SH popup and closing it with ESC key, it seems not to open again...

other than that GREAT WORK, it is a great addition to show the power of this project..

**Comment by axelmohnen** [comment]
Great :-)
@choper725 
Point1: Yes that's right. Unfortunately, with the multi-input control and token approach I can't enter a filter criteria directly in the input field. Entering a value in the multi-input doesn't trigger an event in order to map the input value into token.
@oblomov-dev: The multi-input event tokenUpdate gets triggered, but the event parameters "addedTokens" and "removedTokens" are not forwarded. Any idea how to solve this issue?

**Comment by choper725** [comment]
try passing the event parameter like:
‘${$parameters>/}’

when you trigger the event it will stop at chrome debug, and check what you have in the attributes[1] property

**Comment by oblomov-dev** [comment]
yes i had the same problems with the events, the model gets not updated. Thats why i switched to an approach with a combination of a F4 Help and Popup like here:
Z2UI5_CL_DEMO_APP_056
But i think this is what you are also doing, no better ideas at the moment.

**Comment by choper725** [comment]
added change event to multi_input in order to trigger event upon trying without a token on a field..

**Comment by axelmohnen** [comment]
FYI
I just create a new PR. The SHLP generator "z2ui5_cl_demo_app_104" supports collective searchhelps as well now.

I will test the "change" event asap...

**Comment by choper725** [comment]
error on SH of the the multi_input field.
![image](https://github.com/abap2UI5/abap2UI5/assets/129428114/2e3a06b4-cbff-4df5-bd02-76382ac79af3)


BTW on version 7.5 i get an error on line 461 in the demo here
https://github.com/abap2UI5/abap2UI5-samples/commit/8a6fbbad976927177dcea8f33fbd3a19f87e3252#diff-39e579a5be8b45be7b98975238a05163ac268e4a7359d374cf3772b20aa59da8R461

**Comment by axelmohnen** [comment]
@choper725 :
Unfortunately, I can't re-produce your error. Are you using the BP searchhelp "BUPAP" from the demo?

**Comment by choper725** [comment]
no, im using ddic type collective shlp.

and regarding the secons issue, the select from internal table is not supported in abap < 7.52

**Comment by oblomov-dev** [comment]
probably the error occurs because @axelmohnen is using a certain tag of abap2UI5 and @choper725 uses the main branch. the binding logic changed slightly. @choper725 try to replace the method _bind with _bind_local in the example, that fix helped me last time.

**Comment by axelmohnen** [comment]
Regarding the "change" event on multi-input:
The event gets nicely triggered and the value is correctly binded to the "VALUE" property of the multi-input control.
So, theoretically at that point I could convert the user input into a multi-input token, but deleting a value or token triggers the event as well, but I don't have an identifier at that moment. I can't differentiate between "add" and "remove".
Any ideas are welcome like usually :-)

@choper725 : 
the T_ARG parameter (‘${$parameters>/}’) didn't work. Could you please explain me more in detail how to do this?

Thanks a lot.

**Comment by choper725** [comment]
@axelmohnen ill take a look at this

**Comment by axelmohnen** [comment]
Hi,

I found a way to retrieve the "deleted" token information via event "tokenUpdate".
I'm not sure if this is a clean solution, but it works ;-)
![image](https://github.com/abap2UI5/abap2UI5/assets/22371169/a3071406-d003-4f75-a074-0b3479400dff)
![image](https://github.com/abap2UI5/abap2UI5/assets/22371169/ab5a63fe-5196-4050-8339-d65b8bb6a4df)

**Comment by oblomov-dev** [comment]
Cool, never checked removed tokens, looks like a good idea!

btw. "I'm not sure if this is a clean solution, but it works... "
this is a good sentence describing the whole project ;)

**Comment by axelmohnen** [comment]
FYI
I just create a PR for DDIC SHLP generator app "Z2UI5_CL_DEMO_APP_104".
It contains some cleanups and new features
- Manual creation and deletion of tokens via input field
- 2 user exits for prefilling select-option and filtering the search result. 
Therefore I introducing two new optional import parameters (IV_RESULT_FILTER_EXIT and IV_SELOPT_PREFILL_EXIT)
in the FACTORY() method.  For more details, please have a look to my comments in the FACTORY() method.

Open Point:
Shall we keep the generator app in the demo repo or shall we create a separated repo for utility apps?

**Comment by oblomov-dev** [comment]
thank you for your work!

Yes i think the demo repository is a good place only for code samples but this is a complete functionality for reuse so maybe not the best place. To keep the abap2UI5 framework small i would also not merge this to the main repository, so a utility repository is a good idea. I collected helper functionalities which were demos before already here:
https://github.com/abap2UI5/ext-app_tools
Maybe this could also be the right place for your generic search help?

**Comment by axelmohnen** [comment]
Yes, that's a great idea.
Here is a proposal for the name "ZCL_2UI5_TOOL_APP_SHLP_GEN".

**Comment by oblomov-dev** [comment]
ok reorganized it now. the classes 102, 104,105 are in abap2UI5-tools now:
<img width="1164" alt="image" src="https://github.com/abap2UI5/abap2UI5/assets/102328295/07933aac-d24c-4773-aa30-8bc4881117d1">

https://github.com/abap2UI5/abap2UI5-tools

I also renamed them that they fit to the other classes in this package. Need to do some more cleanup, but for the beginning i hope everything works and is as you expected. Feel also free to change everything and again thanks for the contribution!

**Comment by choper725** [comment]
seems to work GREAT @axelmohnen !
just two issues with it,
1. when closing the SH popup with ESC key, seems not want to repoen again..
2. when using collective SH like, DD_TYPES.  input fields for SH of view and table types not appearing..

btw i changed the select from internal table to read table, for downport, @axelmohnen tell me if its fine by you to PR this into the generic_SH class..

**Comment by axelmohnen** [comment]
@choper725 
1) I could re-produce the error, I will check this. It seems the whole screen gets freezed.
2) Can't re-produce. For it works out-of-the-box. Do you get an error message or dump?
![image](https://github.com/abap2UI5/abap2UI5/assets/22371169/ab58c593-9a34-4e5a-bc75-4e84552b0459)

3) PR is fine for me. 
Thanks a lot!

**Comment by choper725** [comment]
press search for view..

should have 2 input fields for this SH

thnx ill do so !
